cmaes/_safe_cma.py

```diff
@@ -0,0 +1,702 @@
+        )
+
+        self._weights = weights
+
+        # evolution path
+        self._p_sigma = np.zeros(n_dim)
+        self._pc = np.zeros(n_dim)
+
+        if cov is None:
+            self._C = np.eye(n_dim)
+        else:
+            assert cov.shape == (n_dim, n_dim), "Invalid shape of covariance matrix"
+            self._C = cov
+
+        self._D: Optional[np.ndarray] = None
+        self._B: Optional[np.ndarray] = None
+
+        self._rng = np.random.RandomState(seed)
+
+        # initial distribution parameter
+        self._sigma = sigma
+        mean, sigma = self._init_distribution(sigma)
+
+        assert sigma > 0, "sigma must be non-zero positive value"
+
+        assert np.all(
+            np.abs(mean) < _MEAN_MAX
+        ), f"Abs of all elements of mean vector must be less than {_MEAN_MAX}"
+
+        self._mean = mean.copy()
+        self._sigma = sigma
+
+        # bounds contains low and high of each parameter.
+        assert bounds is None or _is_valid_bounds(bounds, mean), "invalid bounds"
+        self._bounds = bounds
+        self._n_max_resampling = n_max_resampling
+
+        self._g = 0
+
+        # Termination criteria
+        self._tolx = 1e-12 * sigma
+        self._tolxup = 1e4
+        self._tolfun = 1e-12
+        self._tolconditioncov = 1e14
+
+        self._funhist_term = 10 + math.ceil(30 * n_dim / population_size)
+        self._funhist_values = np.empty(self._funhist_term * 2)
+
+    def _compute_lipschitz_constant(self) -> np.ndarray:
+
+        likelihood = gpytorch.likelihoods.GaussianLikelihood(
+            noise_constraint=gpytorch.constraints.GreaterThan(0)
+        )
+        likelihood.noise = 0
+
+        B, D = self._eigen_decomposition()
+        invSqrtC = cast(np.ndarray, B.dot(np.diag(1 / D)).dot(B.T))
+
+        num_data = int(np.min((len(self.sampled_safe_evals), self.sample_num_lip)))
+        prev_x = self.sampled_points[-num_data:]
+        z_points = (prev_x - self._mean).dot(invSqrtC) / self._sigma
+
+        target_safe_evals = self.sampled_safe_evals[-num_data:]
+        evals_mean = np.mean(target_safe_evals, axis=0)
+        evals_std = np.std(target_safe_evals, axis=0)
+        modified_evals = (target_safe_evals - evals_mean) / evals_std
+
+        # function that returns the norm of gradient
+        def df(x: np.ndarray, model: ExactGPModel) -> torch.Tensor:
+
+            out_scalar = x.ndim == 1
+            x = np.atleast_2d(x)
+
+            grad_norm = torch.zeros(len(x))
+
+            X = torch.autograd.Variable(
+                torch.Tensor(np.atleast_2d(x)), requires_grad=True
+            )
+            mean = likelihood(model(X)).mean
+            dxdmean = torch.autograd.grad(mean.sum(), X)[0]
+
+            grad_norm = torch.sqrt(torch.sum(dxdmean * dxdmean, dim=1))
+
+            if out_scalar:
+                grad_norm = grad_norm.mean().to(torch.float64)
+
+            return -grad_norm
```

Comment:
`df` is defined as `# function that returns the norm of gradient`, but it currently returns a negative value, which seems odd. 
I understand this is intended for `scipy.optimize.minimize`, but it would be better to align it with the description.